src/classes/CW_AudienceGroupController.cls

```diff
@@ -0,0 +1,543 @@
+	{
+		selectAll=false;
+		selectDeselectAllItems();
+		getSelectedCount();
+	}
+	
+	/*								*/
+	/*   NAVIGATION FUNCTIONALITY  	*/
+    /*								*/
+	
+	//go to first page
+	public void first()
+	{
+		this.setGroups.first();
+		getHasFirstAndLast();
+	}
+	
+	//go to last page
+	public void last()
+	{
+		this.setGroups.last();
+		getHasFirstAndLast();
+	}
+	
+	//advance to next page
+	public void doNext()
+	{
+		if(this.setGroups.getHasNext())
+		{
+			this.setGroups.next();
+			getHasFirstAndLast();
+			resetCheckBoxes();
+		}
+	}
+
+ 	//advance to previous page
+ 	public void doPrevious()
+ 	{
+		if(this.setGroups.getHasPrevious())
+		{
+			this.setGroups.previous();
+			getHasFirstAndLast();
+			resetCheckBoxes();
+		}
+	}
+	
+	//check whether first and last page exists
+	public void getHasFirstAndLast()
+	{
+		if (getPageNumber()==1 && getPageNumber() != getTotalPages())
+		{
+			hasFirstPage = false;
+			hasLastPage = true;
+		} 
+		else if (getPageNumber() == getTotalPages() && getTotalPages()!=1)
+		{
+			hasFirstPage = true;
+			hasLastPage = false;
+		} 
+		else if (getPageNumber() == getTotalPages() && getTotalPages()==1)
+		{
+			hasFirstPage = false;
+			hasLastPage = false;
+		}
+		else {
+			hasFirstPage = true;
+			hasLastPage = true;
+		} 
+	}
+
+	//return whether previous page exists
+	public Boolean getHasPrevious()
+	{
+		if(this.setGroups!=null)
+			return this.setGroups.getHasPrevious();
+		else 
+			return null;
+	}
+
+	//return whether next page exists
+	public Boolean getHasNext()
+	{
+		if(this.setGroups!=null)
+			return this.setGroups.getHasNext();
+		else 
+			return null;
+	}
+
+	//return page number
+	public Integer getPageNumber()
+	{
+		if(this.setGroups!=null)
+			return this.setGroups.getPageNumber();
+		else 
+			return null;
+	}
+
+	//return total pages
+	Public Integer getTotalPages()
+	{
+		if (this.setGroups!=null)
+		{
+			pageSize = this.setGroups.getPageSize();
+			pages = totalSize/pageSize;
+		
+			return (Integer)pages.round(System.RoundingMode.CEILING);
+		}
+		else 
+			return null;
+	}
+	
+	//return count of selected items
+	public Integer getSelectedCount()
+	{
+		return this.selectedGroupIds.size();
+	}
+	
+	//return a set of control groups
+	public set<Id> getControlGroupIds()
+	{
+		controlGroupsIds = new Set<Id>();
+						  
+		for (CW_Audience_Group__c curr : audienceGroups)
+		{
+			if (curr.Control_Group__c!=null)
+				controlGroupsIds.add(curr.Control_Group__c);
+		}
+		
+		return controlGroupsIds;
+	}
+	
+	public String getDeleteMessage()
+	{
+		List<CW_Audience_Group__c> groupsToDelete = new List<CW_Audience_Group__c>();
+		List<CW_Audience_Group__c> groupsSelected = new List<CW_Audience_Group__c>();
+		List<CW_Audience_Group__c> selectedControlGroups = new List<CW_Audience_Group__c>();
+		
+		//get the set of control groups
+		controlGroupsIds = getControlGroupIds();
+		//list of groups selected for deletion
+		groupsSelected = [SELECT Id, Name, Is_Control_Group__c, 
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.